### PR TITLE
[MemCpyOpt] Fix infinite loop in memset+memcpy fold

### LINKED ISSUE #98610: Clang hangs forever building under -O1
ToT Clang hangs when building under -O1.

```
clang++ test.i -fno-exceptions -O1 -c -emit-llvm
```

The reproducer, `test.i`:

```c++
class MockDevice;
struct sp {
  MockDevice operator*();
};
void memset(void *__dest, int __ch, unsigned __len) {
  long __trans_tmp_1 = __builtin_object_size(__dest, 0);
  __builtin___memset_chk(__dest, __ch, __len, __trans_tmp_1);
}
template <class>
class function;
template <class _Rp, class... _ArgTypes>
class function<_Rp(_ArgTypes...)> {
 public:
  template <class _Fp>
  function operator=(_Fp);
};
template <typename T>
struct hidl_vec {
  hidl_vec() : mSize() { memset(mPad, 0, 0); }
  hidl_vec(hidl_vec &other) { *this = other; }
  hidl_vec &operator=(hidl_vec &other) {
    copyFrom(other, other.mSize);
    return *this;
  }
  T &operator[](unsigned index) { return mBuffer[index]; }
  T* mBuffer;
  unsigned mSize;
  unsigned mPad[];
  template <typename Array>
  void copyFrom(Array &data, unsigned size) {
    mSize = size;
    if (mSize) mBuffer = new T();
    for (int i = 0; i < size; ++i) mBuffer[i] = data[i];
  }
};
struct Capabilities {
  struct OperandPerformance {
    int a, b, c;
  };
  hidl_vec<OperandPerformance> operandPerformance;
};
template <typename>
class Action;
template <typename R, typename... Args>
class Action<R(Args...)> {
  using F = R();
 public:
  template <typename G>
  Action(G fun) {
    fun_ = fun;
  }
  function<F> fun_;
};
struct OnCallSpec {
  OnCallSpec WillByDefault(Action<void()>);
};
struct MockSpec {
  OnCallSpec InternalDefaultActionSetAt();
};
struct MockDevice {
  MockSpec gmock_getCapabilities();
};
sp createMockDevice_mockDevice;
MockDevice createMockDevice() {
  Capabilities getCapabilities_ret;
  (*createMockDevice_mockDevice)
      .gmock_getCapabilities()
      .InternalDefaultActionSetAt()
      .WillByDefault(getCapabilities_ret);
  return *createMockDevice_mockDevice;
}
```

### PR DESCRIPTION
**Pull request by nikic**
For the case where the memcpy size is zero, this transform is a complex no-op. This can lead to an infinite loop when the size is zero in a way that BasicAA understands, because it can still understand that dst and dst + src_size are MustAlias.

I've tried to mitigate this before using the isZeroSize() check, but we can hit cases where InstSimplify doesn't understand that the size is zero, but BasicAA does.

As such, this bites the bullet and adds an explicit isKnownNonZero() check to guard against no-op transforms.

Fixes https://github.com/llvm/llvm-project/issues/98610.

### CODE REVIEW

**Comment by nikic** [comment]
For the record, https://github.com/llvm/llvm-project/commit/07460b66666403fc86e01d3fed6b77c1baf94cb1 was my previous attempt to fix this.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/test/Transforms/MemCpyOpt/memcpy-zero-size.ll [hunk 1]
```diff
@@ -34,3 +34,18 @@ define void @pr64886(i64 %len, ptr noalias %p) {
   call void @llvm.memcpy.p0.p0.i64(ptr inttoptr (i64 -1 to ptr), ptr %p, i64 poison, i1 false)
   ret void
 }
+
+define void @pr98610(ptr %p, ptr noalias %p2) {
```
Comment (antoniofrighetto) [inline llvm/test/Transforms/MemCpyOpt/memcpy-zero-size.ll:38]: Hence, it is irrelevant which is the `noalias` pointer in the context of this issue, right?

**Review by dianqk — APPROVED** [review]
LGTM

**Comment by kongy** [comment]
Verified that this fixed https://github.com/llvm/llvm-project/issues/98610 in its original (non-reduced) code.

**Pull request merged**